### PR TITLE
kubeadm: Fix an upgrading issue wrt the dynamic kubelet dropin

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
At the moment (will change), this PR does two things:
 - Makes `kubeletphase.WriteKubeletDynamicEnvFile` run at upgrade time if needed for making v1.10 -> v1.11 upgrades work end to end
 - Fixes so that `--dry-run` works smoothly on `init/upgrade` ==> separate PR

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/kind bug
/priority critical-urgent
/status approved-for-milestone
/assign @timothysc

### CODE REVIEW

**Comment by luxas** [comment]
/hold
I'll split the two distinct changes currently included in this PR into two PRs.

File: cmd/kubeadm/app/phases/upgrade/postupgrade.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/upgrade/postupgrade.go]: why is this not in the dryRun section?

**Review by timothysc — COMMENTED** [review]
comments need clarification.

I'm starting to not become a fan of how dryRun percolates across signatures when it's a global knob.  It makes more sense to have it be part of the global config.  Otherwise we add 0 value mundging all the signature to switch on a bool. 

I think at this stage it's too much risk to make that change, but I think we should open an issue for 1.12.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by dims** [comment]
/lgtm

re-apply lgtm after a rebase

**Pull request merged**
